[Coverage] Improve performance of propagating Counter of Expansions

chapuni opened the pull request:
This improves `clang++ -dump-coverage-mapping` (939,498 lines) for `RISCVInstructionSelector.cpp` from 30m to 1m18s and also improves `llvm-cov` for `check-llvm` from 33m to 24s.

The current implementation behaved O(N^2) order with hundreds thousands of Expansions.

This obsoletes modifying Counters on Expansions and calculates Counter values in the Counter evaluation with traversal of Expansions.

Depends on #125407

Fixes #113173

chapuni committed:
Update desc

chapuni committed:
Merge branch 'main' into users/chapuni/cov/binary

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test
	llvm/test/tools/llvm-cov/showLineExecutionCounts.test

chapuni committed:
Merge branch 'users/chapuni/cov/binary' into users/chapuni/cov/single/refactor

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/getpair

chapuni committed:
Introduce BranchCounterPair

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/pair

chapuni committed:
void verifyCounterMap() const

chapuni committed:
Introduce the dedicated class CounterPair instead of std::pair

chapuni committed:
Merge branches 'users/chapuni/cov/single/getpair' and 'users/chapuni/cov/single/pair' into users/chapuni/cov/single/nextcount-base

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CodeGenPGO.cpp
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Introduce {UseExecPath, UseSkipPath} instead of {false, true}

chapuni committed:
Catch up the merge

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Eliminate (rewind) "refactor" changes

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/switch

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/if

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/loop' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/condop' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop' into users/chapuni/cov/single/trunk

chapuni committed:
Fix wrong merge resolutions

chapuni committed:
Reorganize CoverageMapping::SingleByteCoverage

chapuni committed:
Prune commented-out line

chapuni committed:
[Coverage] Make `MCDCRecord::Folded` as `[false/true]` with BitVector. NFC.

chapuni committed:
[Coverage] MCDC: Move `findIndependencePairs` into `MCDCRecord`

chapuni committed:
Merge branches 'users/chapuni/cov/merge/mcdcfold' and 'users/chapuni/cov/merge/mov_ind' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
[Coverage] Sort `MCDCRecord::ExecVectors` order by Bitmap index

chapuni committed:
llvm-cov: Refactor CoverageSummaryInfo. NFC.

chapuni committed:
Merge remote-tracking branches 'origin/users/chapuni/cov/single/refactor' and 'origin/users/chapuni/cov/single/unify' into users/chapuni/cov/merge/forfile-base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/refactor

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test
	llvm/test/tools/llvm-cov/showLineExecutionCounts.test
	llvm/tools/llvm-cov/CodeCoverage.cpp
	llvm/tools/llvm-cov/SourceCoverageView.h

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/refactor

chapuni committed:
Merge branch 'users/chapuni/cov/single/refactor' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'users/chapuni/cov/single/refactor' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/trunk

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/summaryinfo

chapuni committed:
Merge branch 'users/chapuni/cov/merge/summaryinfo' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/switch

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/if

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/binop-base

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/loop' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/condop' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

Conflicts:
	llvm/test/tools/llvm-cov/branch-macros.test

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branches 'users/chapuni/cov/single/switch' and 'users/chapuni/cov/single/binop' into users/chapuni/cov/single/trunk

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp

chapuni committed:
Add a test

chapuni committed:
Prune redundant logic

chapuni committed:
Merge branch 'users/chapuni/cov/single/replace' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

chapuni committed:
Expand RHS of MapToExpand. This will prevent recursion.

chapuni committed:
Append an explanation in the comment

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/binop-base

chapuni committed:
Update test

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mov_ind

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'users/chapuni/cov/single/unify' into users/chapuni/cov/merge/mcdcsort-base

chapuni committed:
Merge branch 'users/chapuni/cov/merge/mcdcsort-base' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'users/chapuni/cov/merge/mov_ind' into users/chapuni/cov/single/unify

chapuni committed:
Revert "Merge branch 'users/chapuni/cov/single/unify' into users/chapuni/cov/merge/mcdcsort-base"

This reverts commit c33e8987289f2775792977ee279c4e458f8091d5, reversing
changes made to ad6726dbfac5074c856f41adadb3b5bc14580311.

chapuni committed:
Rewind switch DefaultCase. (to #113112)

chapuni committed:
Enable addCounters

chapuni committed:
Get rid of structual bindings

chapuni committed:
Flatten with getBranchCounterPair(SkipCntForOld)

chapuni committed:
Merge branch 'users/chapuni/cov/single/getpair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/switch

chapuni committed:
Resurrect merge conflict

chapuni committed:
Revert "Rewind switch DefaultCase. (to #113112)"

This reverts commit d854fb123be0449e83093f11e9f826c9baa3b766.

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

chapuni committed:
Dissolve assert

chapuni committed:
Dissolve assert

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/if

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/binop-base

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branches 'users/chapuni/cov/single/loop', 'users/chapuni/cov/single/condop' and 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/trunk

chapuni committed:
Update comments

chapuni committed:
Decorate the mock

chapuni committed:
Dissolve the old behavior

chapuni committed:
Will be pruned after the migration of SingleByte.

chapuni committed:
Add comments

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/replace

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/getpair

chapuni committed:
Merge branch 'users/chapuni/cov/single/replace' into users/chapuni/cov/single/getpair

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/pair

chapuni committed:
Merge branch 'users/chapuni/cov/single/getpair' into users/chapuni/cov/single/pair

chapuni committed:
Merge branch 'users/chapuni/cov/single/pair' into users/chapuni/cov/single/nextcount-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount-base' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/base

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/condop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/if

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/base' into users/chapuni/cov/single/switch

chapuni committed:
Merge branches 'users/chapuni/cov/single/loop', 'users/chapuni/cov/single/condop' and 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop' into users/chapuni/cov/single/trunk

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
[Coverage] Improve performance of propagating Counter of Expansions

This improves `-dump-coverage-mapping` (939,498 lines) for
`RISCVInstructionSelector.cpp` from 30m to 1m18s and also improves
`llvm-cov` for `check-llvm` from 33m to 24s.

The current implementation behaved O(N^2) order with hundreds
thousands of Expansions.

This assumes:
  - Records are partitioned by FileID.
    - ExpandedFileID doesn't point FileID==0, since it is the root.
  - The Count in Expansion is propagated from 1st Record in
    ExpandedFileID.

Therefore another fact below can be assumed.
  - Propagation chain consists of Expansions at each 1st Record.

This scans the Record at most a few times. O(N) is expected.

Fixes #113173

chapuni committed:
Get rid of the assumption "[0] is the root"

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:
```diff
@@ -440,26 +441,83 @@ Error RawCoverageMappingReader::read() {
-        FileIDExpansionRegionMapping[R.FileID] = nullptr;
-      }
+
+    if (PrevFileID != R.FileID) {
```
We can reduce indentation here.

```
// Have we found the first record for this file already?
if (PrevFileID == R.FileID) {
  // We've seen this FileID before, so we have found the
  // first record. Move on to the next region.
  continue;
}
// We have not. Mark the current record as the first record
// for this file.
assert(!FileIDExpansionRegionMapping[R.FileID].FirstR && "Already found a record for an ID we haven't seen before?!");
FileIDExpansionRegionMapping[R.FileID].FirstR = &R;
// Remember that we've processed this file already.
PrevFileID = R.FileID;
```

chapuni replied:
I dissolved the condition `PrevFileID != R.FileID` to invert the loop by descending order. `FirstR` is updated unconditionally.

I'm still unsure "early return" could help readibility here.

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:
```diff
@@ -440,26 +441,83 @@ Error RawCoverageMappingReader::read() {
+  for (auto &R : MappingRegions) {
+    if (R.Kind == CounterMappingRegion::ExpansionRegion) {
+      // The File that contains Expansion may be a node.
+      assert(!FileIDExpansionRegionMapping[R.ExpandedFileID].ExpanderR);
```
Is this checking that there is only one expansion region?

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:
```diff
@@ -440,26 +441,83 @@ Error RawCoverageMappingReader::read() {
+    if (R.Kind == CounterMappingRegion::ExpansionRegion) {
+      // The File that contains Expansion may be a node.
+      assert(!FileIDExpansionRegionMapping[R.ExpandedFileID].ExpanderR);
+      FileIDExpansionRegionMapping[R.ExpandedFileID].ExpanderR = &R;
```
Might be slightly cleaner to pull this out into a reference variable?

E.g.

```
FileNode &N = FileIDExpansionRegionMapping[R.ExpandedFileID];
assert(!N.ExpanderR && "Already found an expansion region for file node?!");
N.ExpanderR = &R;
N.ExpansionPending = true;
```

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:
```diff
@@ -440,26 +441,83 @@ Error RawCoverageMappingReader::read() {
+    /// The Expansion Record out of this File.
+    CounterMappingRegion *ExpanderR = nullptr;
+    /// Count hasn't been propagated yet (for Expansion)
+    bool ExpansionPending = false;
```
```suggestion
    bool ExpansionCountPropagated = false;
```
Maybe more descriptive?

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:
```diff
@@ -440,26 +441,83 @@ Error RawCoverageMappingReader::read() {
+    /// Count hasn't been propagated yet (for Expansion)
+    bool ExpansionPending = false;
+  };
+  SmallVector<FileNode> FileIDExpansionRegionMapping(VirtualFileMapping.size());
```
```suggestion
  // We will construct a tree with one node per file in the virtual file mapping.
  SmallVector<FileNode> FileIDExpansionRegionMapping(VirtualFileMapping.size());
```
Add a comment explaining what we're going to do with this.

ornata replied:
Also what will the edges in the tree be? That's not clear to me right now.

ornata commented on llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp:
```diff
@@ -440,26 +441,83 @@ Error RawCoverageMappingReader::read() {
   }
 
-  return Error::success();
+  auto Propagate = [&](FileNode &X) {
```
Why a lambda and not a function?

chapuni replied:
Why not a lambda? I prefer using a lambda as definition of subfunction that shares the context.

I could abandon serpentine seek just to get rid of the lambda. I don't want to introduce another method here.

chapuni committed:
[MC/DC] Update CoverageMapping tests

To resolve the error, rename mcdc-error-nests.cpp -> mcdc-nested-expr.cpp at first.

- `func_condop`
  A corner case that contains close decisions.
- `func_expect`
  Uses `__builtin_expect`. (#124565)
- `func_lnot`
  Contains logical not(s) `!` among MC/DC binary operators. (#124563)

mcdc-single-cond.cpp is for #95336.

chapuni committed:
[MC/DC] Make covmap tolerant of nested Decisions

CoverageMappingWriter reorders `Region`s by `endLoc DESC` to
prioritize wider `Decision` with the same `startLoc`.

In `llvm-cov`, tweak seeking Decisions by reversal order to find
smaller Decision first.

chapuni committed:
[MC/DC] Refactor MCDC::State::Decision. NFC.

Introduce `ID` and `InvalidID`. Then `DecisionByStmt` can have three
states.

* Not assigned if the Stmt(Expr) doesn't exist.
* When `DecisionByStmt[Expr]` exists:
  * Invalid and should be ignored if `ID == Invalid`.
  * Valid if `ID != Invalid`. Other member will be filled in the
    Mapper.

chapuni committed:
[MC/DC] Refactor MCDCCoverageBuilder. NFC.

- Get rid of the old `DecisionStack` and dissolve it into push/pop
  `CurCondIDs` in `VisitBin`, since `VisitBin` is recursive.

- Introduce the new `DecisionStack` with `DecisionState` to handle the
  current `Decision` in nested `Decision`s.
  - The stack has the sentinel that has `DecisionExpr = nullptr`.
  - Split out `checkDecisionRootOrPush` from `pushAndAssignIDs` for
    non-BinOp. It assigns `CondID` to `E` (instead of assignment LHS
    in `pushAndAssignIDs`).
  - The stack is manupilated at the top Decision operator in `VisitBin`.
    - The stack grows at the entrance of the Decision with the initial
      state.
    - In the same level in `VisitBin`, the stack is popped and the
      `Decision` record is emitted.
- Introduce `DecisionEndToSince` to sweep `MCDCBranch`es partially in
  `cancelDecision`.

chapuni committed:
[MC/DC] Prune MCDCLogOpStack and use CGF.isMCDCDecisionExpr

`MCDCLogOpStack` is used only for detection of the Decision root. It
can be detected with `MCDC::State::DecisionByStmt`.

chapuni committed:
[MC/DC] Create dedicated MCDCCondBitmapAddr for each Decision

MCDCCondBitmapAddr is moved from `CodeGenFunction` into `MCDCState`
and created for each Decision.

In `maybeCreateMCDCCondBitmap`, Allocate bitmaps for all valid
Decisions and emit them order by ID, to prevent nondeterminism.

chapuni committed:
Merge branches 'users/chapuni/mcdc/nest/tests', 'users/chapuni/mcdc/nest/covmapdesc', 'users/chapuni/mcdc/nest/covgen' and 'users/chapuni/mcdc/nest/logopstack' into users/chapuni/mcdc/nest/nest-base

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/bitmapaddr' into users/chapuni/mcdc/nest/nest-base

chapuni committed:
[MC/DC] Enable nested expressions

A warning "contains an operation with a nested boolean expression." is
no longer emitter. At the moment, split expressions are treated as
individual Decisions.

chapuni committed:
[MC/DC] Handle __builtin_expect as if parenthses

Fixes #124565

chapuni committed:
[MC/DC] Enable usage of `!` among `&&` and `||`

In the current implementation, `!(a || b) && c` was not treated as one
Decision with three terms.

Fixes #124563

chapuni committed:
Update ReleaseNotes

chapuni committed:
Update ReleaseNotes

chapuni committed:
isValid()

chapuni committed:
Add comment

chapuni committed:
Prune "nested" warning.

chapuni committed:
Remove mcdc-single-cond.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/tests

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Prune mcdc-single-cond.cpp

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/tests' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/bitmapaddr' into users/chapuni/mcdc/nest/nest-base

Conflicts:
	clang/test/CoverageMapping/mcdc-single-cond.cpp

chapuni committed:
Dissolve one condition to invert the loop

chapuni committed:
Refactor with expansion walking

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covmapdesc' into users/chapuni/cov/expansion

chapuni committed:
Revise w/ expansion walker

chapuni commented on llvm/unittests/ProfileData/CoverageMappingTest.cpp:
```diff
@@ -707,13 +707,16 @@ TEST_P(CoverageMappingTest, expansion_gets_first_counter) {
+  ProfileWriter.addRecord({"func", 0x1234, {1, 2, 4}}, Err);
+  EXPECT_THAT_ERROR(loadCoverageMapping(), Succeeded());
+
+  auto FunctionRecords = LoadedCoverage->getCoveredFunctions();
+  ASSERT_EQ(1u, std::distance(FunctionRecords.begin(), FunctionRecords.end()));
 
-  ASSERT_EQ(CounterMappingRegion::ExpansionRegion, Output.Regions[2].Kind);
-  ASSERT_EQ(Counter::getCounter(2), Output.Regions[2].Count);
-  ASSERT_EQ(3U, Output.Regions[2].LineStart);
+  const auto &CR = (*FunctionRecords.begin()).CountedRegions;
+  ASSERT_EQ(CounterMappingRegion::ExpansionRegion, CR[2].Kind);
+  ASSERT_EQ(4u, CR[2].ExecutionCount);
+  ASSERT_EQ(3U, CR[2].LineStart);
```
This should work w/o this change. May I split out or migrate this in advance?

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/nextcount

Conflicts:
	clang/lib/CodeGen/CodeGenFunction.h

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/loop

Conflicts:
	clang/lib/CodeGen/CGStmt.cpp
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/condop

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/if

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/switch

chapuni committed:
Merge branches 'users/chapuni/cov/single/loop', 'users/chapuni/cov/single/condop' and 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

chapuni committed:
Merge branches 'users/chapuni/cov/single/binop' and 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/covmapdesc

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/logopstack

chapuni committed:
clangCodeGen: Follow #142155

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/covgen

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp (#172557)

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/bitmapaddr

Conflicts:
	clang/lib/CodeGen/CodeGenFunction.h (#142155)

chapuni committed:
Merge branches 'users/chapuni/mcdc/nest/covmapdesc', 'users/chapuni/mcdc/nest/covgen', 'users/chapuni/mcdc/nest/logopstack' and 'users/chapuni/mcdc/nest/bitmapaddr' into HEAD

chapuni committed:
Merge commit '64cbe975bf1a9425a5bfad0e03f9e0ea26769979' into users/chapuni/mcdc/nest/nest-base

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/nest-base' into users/chapuni/mcdc/nest/nest

Conflicts:
	clang/lib/CodeGen/CodeGenPGO.cpp (#172557)
	clang/test/Frontend/custom-diag-werror-interaction.c

chapuni committed:
Prune warn_pgo_nested_boolean_expr

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Update mcdc-export-json.test (in #159119)

chapuni committed:
-EnableSingleByteCoverage

chapuni committed:
Add comments

chapuni committed:
ReleaseNotes.rst

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covmapdesc' into users/chapuni/cov/expansion

chapuni committed:
emitCounterSetOrIncrement: Simplify.

- s/`TheMap`/`TheCounterPair`/
- Change the type of `Counter` to `ValueOpt`.

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/nextcount

chapuni committed:
Merge branch 'users/chapuni/cov/single/nextcount' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/unify

chapuni committed:
Merge branch 'users/chapuni/cov/single/unify' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'main' into users/chapuni/cov/merge/mcdcsort

chapuni committed:
Merge branch 'users/chapuni/cov/merge/mcdcsort' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/mcdcstate

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/mcdcstate' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/loop

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/condop

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/if

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/switch

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/cov/single/trunk

Conflicts:
	clang/docs/ReleaseNotes.rst

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
fix: LoopFalse should be used.

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/expect

Conflicts:
	clang/include/clang/AST/IgnoreExpr.h
	clang/lib/CodeGen/CodeGenFunction.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/covgen

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/covgen

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/lnot

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/lnot' into users/chapuni/mcdc/nest/logopstack

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/logopstack

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/logopstack' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/expect

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/expect' into users/chapuni/mcdc/nest/covgen

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/covgen

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covgen' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/bitmapaddr' into users/chapuni/mcdc/nest/covmapdesc

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covmapdesc' into users/chapuni/mcdc/nest/nest

Conflicts:
	clang/lib/CodeGen/CodeGenPGO.cpp

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/bitmapaddr' into users/chapuni/mcdc/nest/covmapdesc

chapuni committed:
[Coverage] Introduce the predicate `hasSkipCounter(S)`. NFC.

Also, rename s/getIsCounterPair/getExecSkipCounterExistence/.

chapuni committed:
Merge commit 'cead042fd7b57f7bb454d4ffe6f0413658a4b069' into users/chapuni/cov/single/loop

chapuni committed:
Use hasSkipCounter

chapuni committed:
Merge commit 'cead042fd7b57f7bb454d4ffe6f0413658a4b069' into users/chapuni/cov/single/if

chapuni committed:
Use hasSkipCounter

chapuni committed:
Merge commit 'cead042fd7b57f7bb454d4ffe6f0413658a4b069' into users/chapuni/cov/single/switch

chapuni committed:
Use hasSkipCounter

chapuni committed:
Merge branch 'users/chapuni/cov/single/loop' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

Conflicts:
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Use hasSkipCounter

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop' into users/chapuni/cov/single/trunk

Conflicts:
	clang/docs/ReleaseNotes.rst

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/bitmapaddr

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/bitmapaddr' into users/chapuni/cov/single/condop

Conflicts:
	clang/lib/CodeGen/CGExprScalar.cpp
	clang/lib/CodeGen/CodeGenPGO.cpp
	clang/lib/CodeGen/CoverageMappingGen.cpp

chapuni committed:
Merge branch 'users/chapuni/cov/single/condop' into users/chapuni/cov/single/execskipexists

chapuni committed:
hasSkipCounter

chapuni committed:
Merge branch 'users/chapuni/cov/single/execskipexists' into users/chapuni/cov/single/loop

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.proftext
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.yaml

chapuni committed:
Merge branch 'users/chapuni/cov/single/execskipexists' into users/chapuni/cov/single/if

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.proftext
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.yaml

chapuni committed:
Merge branch 'users/chapuni/cov/single/execskipexists' into users/chapuni/cov/single/switch

chapuni committed:
getSwitchImplicitDefaultCounterPair: Flatten expressions.

I believe the original form was already readable in this context,
but I rewrote it to match the requested style.

chapuni committed:
Merge branch 'users/chapuni/cov/single/execskipexists' into users/chapuni/cov/single/binop-base

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/binop

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/trunk

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/loop

chapuni committed:
Merge branch 'users/chapuni/cov/single/loop' into users/chapuni/cov/single/switch

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.proftext
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.yaml

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/loop

chapuni committed:
Merge branch 'users/chapuni/cov/single/loop' into users/chapuni/cov/single/switch

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/if

Conflicts:
	clang/test/CoverageMapping/single-byte-counters.cpp
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.proftext
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.yaml
	llvm/test/tools/llvm-cov/Inputs/branch-c-general.c
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.proftext
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.yaml

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop-base' into users/chapuni/cov/single/if

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.proftext
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.yaml
	llvm/test/tools/llvm-cov/Inputs/showLineExecutionCounts-single.yaml

chapuni committed:
Merge branch 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.yaml

chapuni committed:
Merge branch 'users/chapuni/cov/single/trunk' into users/chapuni/cov/single/binop

Conflicts:
	llvm/test/tools/llvm-cov/Inputs/branch-c-general-single.yaml

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/switch

chapuni committed:
Merge branch 'users/chapuni/cov/single/switch' into users/chapuni/cov/single/if

chapuni committed:
Merge branch 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/if

chapuni committed:
Merge branch 'users/chapuni/cov/single/if' into users/chapuni/cov/single/binop

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop' into users/chapuni/mcdc/nest/covmapdesc

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covmapdesc' into users/chapuni/mcdc/nest/nest

Conflicts:
	clang/docs/ReleaseNotes.rst

chapuni committed:
Merge branch 'main' into users/chapuni/cov/single/binop

chapuni committed:
Merge branch 'users/chapuni/cov/single/binop' into users/chapuni/mcdc/nest/covmapdesc

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covmapdesc' into users/chapuni/mcdc/nest/nest

chapuni committed:
Merge branch 'main' into users/chapuni/mcdc/nest/covmapdesc

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/covmapdesc' into users/chapuni/mcdc/nest/nest

chapuni committed:
Merge branch 'users/chapuni/mcdc/nest/nest' into users/chapuni/cov/expansion

chapuni committed:
Merge branch 'main' into users/chapuni/cov/expansion

chapuni committed:
Merge commit 'llvmorg-23-init-13816-g0135cf99f3a2' into users/chapuni/cov/expansion

Conflicts:
	llvm/lib/ProfileData/Coverage/CoverageMappingReader.cpp

MaskRay commented:
Could you squash this into a single commit? The large number of commits make it hard to track the core logic of the PR. (Force pushes are discouraged by some reviewers to preserve review history, but that doesn't apply here)

chapuni commented:
@MaskRay Thanks for taking a look and sorry for inconvenience.

I don't understand why you request a squashed commit. I've prepared a rebased branch.
https://github.com/llvm/llvm-project/compare/users/chapuni/cov/expansion-rebased?expand=1

Since #194996 was merged, I don't have a strong opinion that this should be merged.

MaskRay commented:
Hi, I asked for a squashed commit without a lot of merges because then `git log` will show the purpose clearly. That said, I can also look at the final diff relative to main.

I think the PR description should explain explicitly why the zero counter for `Expansion` in the new -dump-coverage-mapping output is desired.

---

The output from Opus 4.7 still requires validation, but it suggests this change is desirable.

```
● Look at the structure of func:

  void func(void) {       File 0, line:17 → line+5:2 = #0   ← function body counter
    int i = 0;
    MACRO;                Expansion,File 0, ... = 0          ← was #0
                          File 0, line:8 → line+2:2 = 0      ← post-MACRO, dead code
    i = 2;
  }
                          File 1, 3:15 → 3:28 = #0           ← MACRO body (`return; bar()`)
                          File 1, 3:23 → 3:28 = 0            ← `bar()` after `return`, dead

  What an Expansion region actually is

  An Expansion region is a pointer, not a counter. It says: "the source range [line]:3..[line]:8 in File 0 is a macro invocation; its body
  lives in File 1 — go look there for coverage." Clang does not emit a counter increment for the macro-invocation site itself. The
  instrumentation that backs the macro's execution lives entirely inside File 1's regions.

  That's why the on-disk Counter for an Expansion region is always Counter::Zero — there is no real counter to store.

  Why the old code printed = #0

  The old Reader synthesized a counter by copying "the first region of the expanded file" back into Expansion.Count. For MACRO inside func:
  - File 1's first region is File 1, 3:15 → 3:28 = #0.
  - So Expansion.Count got back-patched to #0.

  This looked like the Expansion shared func's counter. It actually doesn't, even though the number coincides:
  - func's #0 is the function-entry counter that clang's instrumentation emits at function entry.
  - File 1's #0 is the same counter, only because MACRO is invoked unconditionally on the path that enters func — so the macro body executes
  exactly as many times as the function entry.

  Look at macros.c:48 in the diff for the cleaner case:
  -  MACRO_2; // Expansion,File 0, ... = #2
  +  MACRO_2; // Expansion,File 0, ... = 0
  There the synthesized counter was #2 (the loop body's counter), not #0. The "shared #0" appearance in func is coincidental to that
  particular control-flow position.
```

```
  Why = 0 is desired

  1. The dump now reflects what's actually stored. -dump-coverage-mapping is a debug view of the raw mapping. The Counter slot for an
  Expansion has always been zero on disk; printing #0 was a lie introduced by the Reader's mutation.
  2. The propagated value isn't lost — it moved to the right place. CountedRegionEmitter::walk() still computes "execution count of an
  Expansion = first region of expanded file" and stores it in ExpansionCounterValues[&Region]. emitCountedRegions() then writes that as the
  Expansion's CountedRegion::ExecutionCount, which is what llvm-cov's reports actually consume. The user-visible coverage numbers are
  unchanged.
  3. It separates parse from evaluate. The Reader's job is to parse bytes into mapping regions. Synthesizing derived counter expressions and
  writing them back into the parsed structure conflated parsing with evaluation, and that conflation is what made the old code O(N²).
  4. The synthesized counter was always derivable. Anyone who needed the Expansion's count could compute it from the first region of the
  expanded file — exactly what the new emitter does. Storing it back into Region.Count was redundant denormalization.
```
